crates/miden-objects/src/transaction/script_builder.rs

```diff
@@ -0,0 +1,314 @@
+    TransactionScriptLibraryError(String),
+    #[error("transaction script error")]
+    TransactionScriptError(#[from] TransactionScriptError),
+}
+
+// SCRIPT BUILDER
+// ================================================================================================
+
+/// A builder for compiling note scripts and transaction scripts with optional library dependencies.
+///
+/// The ScriptBuilder simplifies the process of creating transaction scripts by providing:
+/// - A clean API for adding multiple libraries
+/// - Automatic assembler configuration with all added libraries
+/// - Debug mode support
+///
+/// The typical workflow is:
+/// 1. Create a new ScriptBuilder with debug mode preference
+/// 2. Add any required libraries using `add_library()`
+/// 3. Compile your note script with `compile_note_script()`
+/// 4. Compile your transaction script with `compile_tx_script()`
+///
+/// # Note
+/// For scripts that need access to transaction kernel procedures,
+/// use `ScriptBuilder::with_assembler()` and provide an assembler created with
+/// `TransactionKernel::assembler()` from the miden-lib crate.
+pub struct ScriptBuilder {
+    assembler: Assembler,
+    libraries: Vec<Library>,
+}
+
+impl ScriptBuilder {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new ScriptBuilder with the specified debug mode.
+    ///
+    /// This creates a basic assembler. For transaction scripts that need access to
+    /// transaction kernel procedures, use `ScriptBuilder::with_assembler()` and provide
+    /// an assembler created with `TransactionKernel::assembler()`.
+    ///
+    /// # Arguments
+    /// * `in_debug_mode` - Whether to enable debug mode in the assembler
+    pub fn new(in_debug_mode: bool) -> Self {
+        let assembler = Assembler::default().with_debug_mode(in_debug_mode);
+        Self { assembler, libraries: Vec::new() }
+    }
+
+    /// Creates a new ScriptBuilder with a provided assembler.
+    ///
+    /// This is the recommended constructor when you need access to transaction kernel
+    /// procedures. Pass `TransactionKernel::assembler()` as the assembler parameter.
+    ///
+    /// # Arguments
+    /// * `assembler` - A pre-configured assembler (e.g., from TransactionKernel::assembler())
+    /// * `in_debug_mode` - Whether to enable debug mode
+    pub fn with_assembler(assembler: Assembler, in_debug_mode: bool) -> Self {
+        let assembler = assembler.with_debug_mode(in_debug_mode);
+        Self { assembler, libraries: Vec::new() }
+    }
+
+    // LIBRARY MANAGEMENT
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds multiple libraries to the script builder.
+    ///
+    /// # Arguments
+    /// * `libraries` - Vector of compiled libraries to add to the builder
+    pub fn add_library(&mut self, libraries: Vec<Library>) -> Result<(), ScriptBuilderError> {
```

Comment:
nit: instead of `Vec<Library>` we should be able to do something like `impl Iterator<Item = Library>` or maybe `impl IntoIterator<Item = Library>`.